### DISCUSSION TITLE
Can there be repeats of a keys in a hierarchy of structs in the savegame files?

### DISCUSSION THREAD

**Discussion by michalc** [OpenTTD/OpenTTD#11875] [Q&A] [answered]
In https://github.com/OpenTTD/OpenTTD/blob/master/docs/savegame_format.md there don't seem to be limitations on the keys in structs. Specifically when there are structs of structs, the keys _could_ be repeated in the tree in a well defined way?

But, looking around https://github.com/TrueBrain/OpenTTD-savegame-reader/blob/64e14804e477771d0fd449fc185ac1a8dd8a2fe3/savegame_reader/savegame.py#L52 it looks a bit like there is an assumption that in a chunk, each key in a struct of structs that itself defines a nested struct will only appear once. So I guess... 

1. Am I understanding the code right?
2. If I am understanding, is it reasonable to assume that for the foreseeable future, the format will maintain this property - that a key in a struct of structs that itself defines a struct won't repeat?

Context: wanting to understand the format as best as I can, in order to transform certain things for plotting in https://github.com/michalc/OpenTTDLab (which is very much a work in progress)

**Comment by TrueBrain** [comment]
Your questioning is rather vague, so I am guessing a bit what you are actually asking. So my apology if I miss the point completely :D
Let's start with an example, and we can talk after. My answer will be only about the reader of the header, and not about reading the rest of the content.

```
struct root {
  int a;
  int b;
  struct child1 {
    int c;
  }
  struct child2 {
    int c;
  }
}
```

This is a perfectly valid tree in our savegame. The key-paths in this tree here are always unique. A key-path is defined here as the concatenation of a leaf + all its parents. So `root.child1.c` is a key-path. Not just `c`. And I think this is what your question is about (but I am not really sure). There can be multiple `c`s in leafs; just where they are located in their tree, that has to create an unique path.

What is invalid however, is to have another `int a` in `root`. As now the key would not uniquely identify a field. And this is mandatory. So this is invalid:

```
struct root {
  int a;
  int a;
}
```

That will absolutely never happen. Similar, you can't have another `int c` in any of the child structs.

That said, what you see a lot that we prefix keys in sub-structs anyway. This has more to do how the internals of our code works, or with the history of things, than anything else. For example the Gamelog blob, is full of prefixed fields.

![image](https://github.com/OpenTTD/OpenTTD/assets/1663690/caaf4a00-533e-4898-a3f3-c5fca3f3b96a)

There was absolutely no need for `revision.slver` to be like that; it could have been `slver` just fine.

This is btw how both the Python and Rust implementation read the savegame. They first read the root-table-header. And for every struct there-in, it reads the table-header and assigns it to the key in the root-table-header.

So in the above example, Python does:

`tables["root"], size = self._read_table(reader)`
This creates:

```
{
  "a": ...
  "b": ...
  "child1": ...
  "child2": ...
}
```

After that, `child1` header is read:
`tables[field[2]], sub_size = self._read_table(reader)`

Which assigns the table:

```
{
  "c": ...
}
```

to `tables["root"]["child1"]`. So the root table now looks like:

```
{
  "a": ...
  "b": ...
  "child1": {
    "c": ...
  }
  "child2": ...
}
```

So they are not collapsed or anything (which your question seems to be about).

Hopefully I answered what you are asking for. Otherwise: give examples please :D

**Reply by TrueBrain** [reply]
PS: in our C++ (of the game itself) it is a bit more tricky, as it never actually creates a real tree. There everything remains flat, and headers are just concatenated together. But when reading, it follows the tree to get to the right node. This is purely an optimization, as in our save/load we actually never want to know a single field, but always deal with the whole chunk at once. The "tree" as such is stored in the definition of all the chunks; not in the save/load code itself.

For Python / Rust loaders as I wrote them, and I imagine the work you are doing, you actually want to know a single value. There creating the tree first helps. But it is not required to build the actually tree to walk it as a tree. If that makes sense? :D

**Reply by michalc** [reply]
> Your questioning is rather vague

I know... I struggled with it!

> My answer will be only about the reader of the header, and not about reading the rest of the content.

Yep - that is what I'm concerned with

> The key-paths in this tree here are always unique. 

Yes thought/I would have thought this - hence the sort of surprise in my question

>  So they are not collapsed or anything (which your question seems to be about).

Yeah it sort of felt collapsed...

So I think I was slightly confused in my question, I see from your answer there can be some structure. But..., it's only sort of 1 level deep? The code

```python
tables[field[2]], sub_size = self._read_table(reader)
```
The variable `tables` here is the "top level" `tables`, and never a deeper one?

So yes, it can be:
```python
{
  "a": ...
  "b": ...
  "child1": {
    "c": ...
  }
  "child2": ...
}
```
But the code suggests to me it can't ever be:
```python
{
  "a": ...
  "b": ...
  "child1": {
    "c": {
       "d": {
        ...
      }
    }
  }
  "child2": ...
}
```
because it would assign `d` not on `tables["child1"]["c"]`, but instead on `tables["c"]`?

Edit: Or something like that... I will ponder a bit more on this...

**Reply by michalc** [reply]
Actually:

> `tables[field[2]], sub_size = self._read_table(reader)`
>
> Which assigns the table:
> ```python
> {
>   "c": ...
> }
> ```
> to tables["root"]["child1"]

I don't think it would? It would instead assign to `tables["child1"]`? The reason for this is that `tables` is always the top level `tables` object created in the `read_all_tables` function at https://github.com/TrueBrain/OpenTTD-savegame-reader/blob/64e14804e477771d0fd449fc185ac1a8dd8a2fe3/savegame_reader/savegame.py#L61, rather than anything "deeper" in the structure

**Reply by TrueBrain** [reply, marked as answer]
You are correct, and this is also why deep nesting (kinda) works. All table definitions are stored in `tables` and used when a table is found inside the `root` table (or any nested table therein).

So the implementation does also some form of flattening, but it still very much is a tree.

Where the current implementation breaks is when there are two tables on different levels with the same name. Fully allowed in C++, so something I will address in the Python/Rust implementations soon :)

**Reply by michalc** [reply]
Thank you! Good to know I wasn't barking up the completely wrong tree :-)

**Reply by TrueBrain** [reply]
Pushed an update. As expected, there wasn't an actual issue with the current code, as the only sub-tables with the same name were `common` in vehicles, which were identical. But, it wasn't correct, so it is now fixed :)

It is still a flat list of tables, but they are now named by their path. Was easier to fix that way.
